### ISSUE TITLE
Reduce "restart cost" of sync jobs when they fail

### ISSUE THREAD

**Issue by jechols** [uoregon-libraries/newspaper-curation-app#248]
*This ticket is epic and needs thought. There is at least one task to be done here, but maybe two. Both need a separate ticket.*

We have some iffy network mounts that have given us trouble on and off for years, and it's only gotten worse. The biggest problem at the moment is the "restart cost" of really large file copy jobs.

When a `sync_directory` job kicks off, it assumes nothing about prior runs of the same job. This is intentional: if a job is starting, it needs to verify every file it *thinks* it copied previously, because we don't know anything about **why** it failed.

The problem comes in when we're only able to copy about 500 megs at a time prior to an i/o error. When a job needs to copy a few dozen gigs, this means that we're having to read and verify the same files repeatedly. In some cases, when the network is particularly slow and/or we have a huge job that's 90% done, it can take hours just to start copying files again.

There aren't many good fixes here unless we can figure out our own infra problems, but there may be some ways to improve it all the same:

---

**Updates**: Item 1 was a bust. Item 3 no longer makes any sense after some thought - we would have to do a lot of work to resume a job in this way, given how NCA currently works in terms of these jobs (everything is just sent off to `fileutil.SyncDirectory`)

---

**Update 2**: This is dependent on #177 - the best way to accomplish this ticket is with a list of jobs that all have the same priority within a workflow, such that whatever needs to happen after the sync only happens once everything at this job's priority is done. Dunno if the job list will be prebuilt or created as we sync things, but either way we need some way to not care about the order of files/dirs synced while still caring about them all having to finish before moving on.

Justification: the current setup would require all "mini-sync" jobs be in a sequence such that if one has to retry, the rest just sit waiting. It's entirely possible that having all jobs at the same priority gives us exactly the same results, but it *seems* less likely we'll end up with a huge wait. If there are 1000 jobs, and we go with the current system, a network blip that forces job 10 to retry a dozen times means the *entire line* is waiting 24 hours due to the need to be extra careful with backing off failing jobs. But it's unlikely the network will give us trouble this long. A semi-parallel system would fail a job but then be able to pick up the next one at the same priority. Yes, that could fail, too, all the way down the line, even. But even if all 1000 jobs fail, the next retry will be fairly soon, and all 1000 jobs will be trying to finish. This means a 25-retry full-on-stop-everything failure should still mean that most of the content syncs, and manual job reruns are fairly quick.

---

- <del>When an i/o failure occurs, wait a few seconds and try the file again. For small network "blips" this could make a huge difference. Are we having "blips", as opposed to real outages? No idea, but it's probably worth the hour or so of dev to implement this just in case.</del>
- Consider breaking up these huge sync jobs somehow. One job per file copy would be nuts, but maybe we could do one job per directory or something? Still nuts, but it might be the right choice all the same. First, it reduces restart costs. Second, it reduces the risk of interrupting a job. Right now interrupting a four-hour job is almost a Greek tragedy in terms of manual cleanup. If we had 1000 jobs instead, each taking a few minutes max, the risk of an interruption happening in the first place is lower, and the cost of that interruption is potentially orders of magnitude lower.
- <del>Consider storing copy state in the job. I'm not sure how we'd approach this, though. Copy state could be absurd for jobs that need to copy thousands of files spanning 100+ gigs. But if a nice way to manage this could be figured out, it would certainly mean a much quicker recovery - if we already verified the first 80 gigs were successful, no need to re-verify those next restart.</del>

<del>Option 1 is a no-brainer. 2 has the downside of being a pretty big refactoring, but each copy-verify-delete loop would be very nicely isolated. Option 3 would be really great, but I dunno... the state could be just too much data. And I don't like a single job that's so big it's trying to fake-split itself like this.</del>

**Comment by jechols** [comment]
I'm leaning toward option 2 here. 1 is probably a good idea to just jam in, but I'm not actually sure that's so easy to do without changing our internal `gopkg` fileutil thing, and I don't know that we want such a feature in all the places we use dir syncing (not just NCA). Maybe we do, though... is there a downside to always retrying at least once? Hmm.

Option 3 worries me. The state could be easy to mess up in some way that causes interruptions to actually really break stuff. Say state gets half-written and power goes out or something, I dunno. I also *really* don't like the idea of this huge job just faking everything. I said it before, but it just feels wrong, and in a way that I think could really bite us in the future.

Option 2 is really appealing if it can be done without an absurd amount of work. I'm just not yet sure *how* to implement it. One dir per job doesn't really work: there are top-level files. I guess if it were one dir per job, but using a non-recursive sync? We'd also have to verify things extra carefully: if there's even a single lingering file we somehow missed, that could be disastrous. But the general approach here is very nice, because once a job is done, it's just done. We never need to worry about those files again, they're copied, we're good.

**Comment by jechols** [comment]
Item 1 is implemented. A new ticket is needed once I decide on 2, 3, or neither....

**Comment by jechols** [comment]
Item 1 needs to be removed. It appears to be causing NCA to ... fail, in a way. My suspicion is that the network mount drops, NCA retries a few times, eventually writing to some kind of "phantom" location that goes away. I really don't know what I'm talking about, but we found a sync job that NCA thinks is successful, but which missed about 5000 files and left at least one file in a half-written state. So I think NCA needs to just give up if there's even a single failure, and never silently retry.

**Comment by jechols** [comment]
An option for getting item 2 done without going *too* crazy here: syncdir jobs would sync all files at the dir's top level, then spawn new jobs for sub-dirs.

A huge downside in all situations I can imagine: how do we make a job that doesn't start until after a *large collection* of other jobs are complete?

**Comment by jechols** [comment]
Updated main ticket with info about new approach (workflows which contain a set of jobs).

Also worth noting: after turning off silent retries, we're observing the flurry of failures again. I'm not sure why silently retrying caused errors to get hidden, but whatever was going on, errors were definitely happening and being silenced. And now they're not and we're having to wait days to get files moving...

**Comment by jechols** [comment]
QA for this will be manually pushing to UO's production and ... hoping for the best. (Local dev testing was thorough, but still)
